RecoVertex/PrimaryVertexProducer/interface/WeightedMeanFitter.h

```diff
@@ -0,0 +1,468 @@
+      wx = wx <= precision ? 1. / std::pow(precision, 2) : 1. / std::pow(wx, 2);
+
+      wz = p.second.z();
+      wz = wz <= precision ? 1. / std::pow(precision, 2) : 1. / std::pow(wz, 2);
+
+      x += p.first.x() * wx;
+      y += p.first.y() * wx;
+      z += p.first.z() * wz;
+
+      s_wx += wx;
+      s_wz += wz;
+    }
+
+    if (s_wx == 0. || s_wz == 0.) {
+      edm::LogWarning("WeightedMeanFitter") << "Vertex fitting failed at beginning\n";
+      return TransientVertex(GlobalPoint(0, 0, 0), err, *iclus, 0, 0);
+    }
+
+    x /= s_wx;
+    y /= s_wx;
+    z /= s_wz;
+
+    float old_x, old_y, old_z;
+    float xpull;
+    int niter = 0;
+    float mu = 3.;
+    float err_x, err_y, err_z;
+    err_x = 1. / std::sqrt(s_wx);
+    err_y = 1. / std::sqrt(s_wx);
+    err_z = 1. / std::sqrt(s_wz);
+    float s_err_x = 0., s_err_y = 0., s_err_z = 0.;
+    while ((niter++) < maxIterations) {
+      old_x = x;
+      old_y = y;
+      old_z = z;
+      s_wx = 0;
+      s_wy = 0;
+      s_wz = 0;
+      s2_wx = 0;
+      s2_wy = 0;
+      s2_wz = 0;
+      x = 0;
+      y = 0;
+      z = 0;
+      s_err_x = 0.;
+      s_err_y = 0.;
+      s_err_z = 0.;
+
+      for (const auto& p : points) {
+        wx = p.second.x();
+        wx = wx <= precision ? precision : wx;
+
+        wy = wx * wx + err_y * err_y;
+        wx = wx * wx + err_x * err_x;
+
+        wz = p.second.z();
+        wz = wz <= precision ? precision : wz;
+        wz = wz * wz + err_z * err_z;
+
+        xpull = std::pow((p.first.x() - old_x), 2) / wx;
+        xpull += std::pow((p.first.y() - old_y), 2) / wy;
+        xpull += std::pow((p.first.z() - old_z), 2) / wz;
+        xpull = 1. / (1. + std::exp(-0.5 * ((mu * mu) - xpull)));
+        ndof_x += xpull;
+
+        wx = 1. / wx;
+        wy = 1. / wy;
+        wz = 1. / wz;
+
+        wx *= xpull;
+        wy *= xpull;
+        wz *= xpull;
+
+        x += wx * p.first.x();
+        y += wy * p.first.y();
+        z += wz * p.first.z();
+
+        s_wx += wx;
+        s_wy += wy;
+        s_wz += wz;
+
+        s2_wx += wx * wx;
+        s2_wy += wy * wy;
+        s2_wz += wz * wz;
+
+        s_err_x += wx * pow(p.first.x() - old_x, 2);
+        s_err_y += wy * pow(p.first.y() - old_y, 2);
+        s_err_z += wz * pow(p.first.z() - old_z, 2);
+      }
+      if (s_wx == 0. || s_wy == 0. || s_wz == 0.) {
+        edm::LogWarning("WeightedMeanFitter")
+            << "Vertex fitting failed, either all tracks are outliers or they have a very large error\n";
+        return TransientVertex(GlobalPoint(0, 0, 0), err, *iclus, 0, 0);
+      }
+      x /= s_wx;
+      y /= s_wy;
+      z /= s_wz;
+
+      err_x = std::sqrt(s_err_x / s_wx);
+      err_y = std::sqrt(s_err_y / s_wy);
+      err_z = std::sqrt(s_err_z / s_wz);
+
+      if (std::abs(x - old_x) < (precision / 1.) && std::abs(y - old_y) < (precision / 1.) &&
+          std::abs(z - old_z) < (precision / 1.))
+        break;
+    }
+
+    err(0, 0) = err_x * err_x;
```

Comment:
Again, here the square of the errors are used, therefore the sqrt's computed here above (L436-437) are a useless performance weight